File: p2/p274/HIndex.java
package leetcode.p2.p274;

import java.util.Arrays;

public class HIndex {
    /**
     * 08/27/2022 02:23
     * Time complexity: O(nlogn)
     * Space complexity: O(1)
     */
    public int hIndex(int[] citations) {
        Arrays.sort(citations);
        for (int i = 0; i < citations.length; i++) {
            if (i + 1 > citations[citations.length - 1 - i]) {
                return i;
            }
        }
        return citations.length;
    }

    // TODO O(n) approach
}

File: p2/p274/h_index.py
from typing import List


class Solution:
    """
    Jun 02, 2025 14:42
    Time Complexity: O(nlogn)
    Space Complexity: O(n)
    """

    def hIndex(self, citations: List[int]) -> int:
        sorted_citations = sorted(citations, reverse=True)
        result = 0
        for i in range(len(sorted_citations)):
            if sorted_citations[i] >= i + 1:
                result = max(result, i + 1)

        return result
